### PR TITLE
Make Wallet first-class citizens

### LINKED ISSUE #3784: Make Wallet first-class citizens
Make wallets for which at least one of the owner is in the accounts first-class citizens ; ie. being able to send transactions from then from anywhere in the app && keep their balances up-to-date (thus receiving notification on new transaction on an owned wallet).

### LINKED ISSUE #3849: JS error Transfer error ETH from one account to another from Parity UI
I'm using Parity/v1.5.0-unstable-03db4d037-20161215/x86_64-macos/rustc1.13.0
chrome 54.0.2840.98 (64-bit)

```
Uncaught Error: new BigNumber() not a number: undefined
    at D (http://127.0.0.1:8182/vendor.js:31:5244)
    at http://127.0.0.1:8182/vendor.js:31:10949
    at new t (http://127.0.0.1:8182/vendor.js:31:3151)
    at eval (webpack:///./api/format/input.js?:155:37)
    at Array.forEach (native)
    at inOptions (webpack:///./api/format/input.js?:146:88)
    at Parity.postTransaction (webpack:///./api/rpc/parity/parity.js?:248:128)
    at TransferStore.send (webpack:///./modals/Transfer/store.js?:439:40)
    at Object.eval (webpack:///./modals/Transfer/store.js?:560:6051)
    at N (http://127.0.0.1:8182/vendor.js:16:29328)
```

### PR DESCRIPTION
**Pull request by ngotchac**
This PR fixes some issues (balances not updating, not being able to send tokens from a Wallet), and should close #3784 

This adds a new lookup action to the `Contract` class, which checks for a transaction whether the `from` field corresponds to a Wallet or not. If it does, then try to post the transaction/estimate gas with the correct options, ie. `from` being one of the owner of the wallet that is also one of the user account, etc.
Note that this can be recursive : if a Wallet is owned by another Wallet owner by one of my account, it will correctly set the account as the transaction sender, to the middle Wallet, that will call the execute method on the initial Wallet, which will execute the correct transaction.

The gas estimation is sometimes (more or less 2/3) wrong. So to test one might want to multiply the gas by 2. Maybe this should be taken into account in the PR (while the gas estimation gets fixed) ?

As the modification lies in the Contract class, this also work within dApps ! The only thing not available for Wallets is contract deployment (that would require a modification of the Wallet contract itself)...

Might fix #3849

### CODE REVIEW

**Comment by gavofyork** [comment]
cool - we should probably add a `create` function to our standard wallet contracts. how easily will this play with `DeadMansSwitch`?

**Comment by ngotchac** [comment]
@gavofyork So the DeadManSwitch contract is working fine when backup/owner are Wallets Contracts (I just tested every methods), from the Contracts view. However, the DeadManSwitch contract cannot be used as a Wallet for now. I think we might want to open the Wallet options to other contracts than the _standard_ one. However it would take some work to get there, as we need to make the events and methods that we use more general.

File: js/src/redux/providers/signerReducer.js [hunk 1]
```diff
@@ -90,7 +90,7 @@ export default handleActions({
   signerSuccessRejectRequest (state, action) {
     const { id } = action.payload;
     const rejected = Object.assign(
-      state.pending.find(p => p.id === id),
+      state.pending.find(p => p.id === id) || { id },
```
Comment (jacogr) [inline js/src/redux/providers/signerReducer.js]: Same as the previous fix (now for reject), I missed this case the first time around. (Let's hope it is indeed dead now)

File: js/src/redux/providers/personalReducer.js [hunk 2]
```diff
@@ -41,9 +40,7 @@ export default handleActions({
       contacts,
       hasContacts: Object.keys(contacts).length !== 0,
       contracts,
-      hasContracts: Object.keys(contracts).length !== 0,
-      wallets,
-      hasWallets: Object.keys(wallets).length !== 0
+      hasContracts: Object.keys(contracts).length !== 0
```
Comment (jacogr) [inline js/src/redux/providers/personalReducer.js]: While we are here, could we just fix the ordering as well?

File: js/src/api/contract/contract.js [hunk 3]
```diff
@@ -209,9 +221,93 @@ export default class Contract {
     return `0x${data || ''}${call || ''}`;
   }
 
-  _encodeOptions (func, options, values) {
+  /**
+   * Check whether the given address could be
+   * a Wallet. The result is cached in order not
+   * to make unnecessary calls on non wallet accounts
+   */
+  _isWallet (address) {
```
Comment (jacogr) [inline js/src/api/contract/contract.js]: Not 100% sure this is the correct place for this functionality. Can we walk through today and discuss?

File: js/src/ui/Form/AddressSelect/addressSelect.js [hunk 4]
```diff
@@ -183,6 +182,12 @@ class AddressSelect extends Component {
     }
 
     const id = `addressSelect_${++currentId}`;
+    const ilHint = typeof hint === 'string' || !(hint && hint.props)
+      ? (hint || '')
+      : this.context.intl.formatMessage(
```
Comment (jacogr) [inline js/src/ui/Form/AddressSelect/addressSelect.js]: The issue with this type of formatting is that the message will never be extracted. Hence we always opt for `FormattedMessage` directly (which does get extracted and then becomes available for translation). 

Not sure I understand what causes this issue and how to work around.

File: js/src/views/Signer/components/TransactionPendingForm/TransactionPendingFormConfirm/transactionPendingFormConfirm.js [hunk 5]
```diff
@@ -60,6 +60,7 @@ class TransactionPendingFormConfirm extends Component {
         <Form>
           { keyInput }
           <Input
+            autoFocus
```
Comment (jacogr) [inline js/src/views/Signer/components/TransactionPendingForm/TransactionPendingFormConfirm/transactionPendingFormConfirm.js]: Where does it focus when there are multiple requests? Still on the first?

File: js/src/views/Wallet/wallet.js [hunk 6]
```diff
@@ -64,13 +64,14 @@ class Wallet extends Component {
...
     balance: nullableProptype(PropTypes.object.isRequired),
     images: PropTypes.object.isRequired,
-    address: PropTypes.string.isRequired,
-    wallets: PropTypes.object.isRequired,
+    isTest: PropTypes.bool.isRequired,
+    owned: PropTypes.bool.isRequired,
+    setVisibleAccounts: PropTypes.func.isRequired,
     wallet: PropTypes.object.isRequired,
-    isTest: PropTypes.bool.isRequired
+    walletAccount: nullableProptype(PropTypes.object.isRequired)
```
Comment (jacogr) [inline js/src/views/Wallet/wallet.js]: Possibly `nullableProptype(PropTypes.object).isRequired` (the full property is required, but could be either null or object)

File: js/webpack/vendor.js [hunk 7]
```diff
@@ -64,6 +64,13 @@ module.exports = {
       }
     ]
   },
+
+  resolve: {
+    alias: {
+      '~': path.resolve(__dirname, '../src')
```
Comment (jacogr) [inline js/webpack/vendor.js]: +1

File: js/src/ui/Form/AddressSelect/addressSelect.js [hunk 8]
```diff
@@ -183,6 +182,12 @@ class AddressSelect extends Component {
...
+        hint.props,
+        hint.props.values || []
```
Comment (jacogr) [inline js/src/ui/Form/AddressSelect/addressSelect.js]: `hint.props.values || {}`

**Comment by ngotchac** [comment]
Had to revert a bit the scope of the PR. From now on, the Wallets are hidden in the Dapps Permission modals, which means that they are not available in dApps. This is because it would break dapps using Web3 which doesn't handle Wallet Contracts. This might become possible in 1.6 though.
A solution would be to be able to modify the `from`, `to`, `value` and `data` fields from within the Signer (as it is done for `gas`, `gasPrice` and `minBlock` right now) ; which is not possible ATM.

File: js/src/api/contract/contract.js [hunk 9]
```diff
@@ -14,8 +14,14 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
-import Abi from '../../abi';
+import BigNumber from 'bignumber.js';
+import { intersection } from 'lodash';
 
+import Abi from '~/abi';
+import WalletAbi from '~/contracts/abi/wallet.json';
+import WalletsUtils from '~/util/wallets';
```
Comment (jacogr) [inline js/src/api/contract/contract.js]: As discussed, these two lines above are UI and implementation specific, and doesn't belong in the API. The API layer takes what is provided and encodes that, sends it, decodes the responses, the role is not to make decisions and tweaks on behalf of the caller (caller is either us or other developers).

Changes made here is valid for the app, just needs a different home outside of `~/api`

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 234919d9e92854f69f7d514b0551c1b6dbb9eab5
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by gavofyork** [comment]
Yeah - we'll need an extensible architecture so that code hashes (i.e. types of wallet contracts) can be registered with specialised means of signing. This way dapps can use various addresses with `eth_postTransaction` and the signing UI will still work correctly - either passing it on to a derivative address (like for DeadMansSwitch) or multiple such addresses (the multisig wallet) or something else.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 234919d9e92854f69f7d514b0551c1b6dbb9eab5
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 234919d9e92854f69f7d514b0551c1b6dbb9eab5
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 234919d9e92854f69f7d514b0551c1b6dbb9eab5
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/ethcore/parity

---
This message was auto-generated by https://gitcop.com

**Pull request merged**